src/arch/xtensa/include/arch/atomic.h

```diff
@@ -63,6 +86,25 @@ static inline int32_t arch_atomic_sub(atomic_t *a, int32_t value)
 		: "memory");
 
 	return current;
+#elif XCHAL_HAVE_EXCLUSIVE && CONFIG_XTENSA_EXCLUSIVE && __XCC__
+	/*reference xtos : xipc_misc.h*/
+	int32_t current;
+	int32_t result = 0;
+
+	while (!result) {
+		current = XT_L32EX((int *)a);
+		result = current - value;
+		XT_S32EX(result, (int *)a);
+		XT_GETEX(result);
+	}
+	return current;
+#elif CONFIG_NUM_CORES == 1
+	/* This is fine on unicore  UP systems */
+	*a += value;
+	return *a;
+#else
+	#error No atomic methods available.
```

Comment:
Hi @lgirdwood ,

Do I need to add XCHAL_HAVE_S32C1I?

if we use gcc to compile, it seems compiler will go to "#error No atomic methods available".

Reply:
Yes. we need 
```c
#if CONFIG_LOCK_EXCLUSIVE && __XCC__ && XCHAL_EXCLUSIVE
 /* do this */
#elif XCHAL_S32C1I
/* do that */
#elif CONFIG_NUM_CORES == 1
/* do something else */
#else 
#error No atomic support for multicore system
#endif
```

Reply:
See the CI build
```
warning: the int symbol CORE_COUNT (defined at src/platform/Kconfig:272) has a non-int default MP_NUM_CPUS (undefined)
-- MEU_OFFSET=1344
-- Configuring done
-- Generating done
-- Build files have been written to: /srv/home/jenkins/workspace/sof_generic_build/build_tgl_xcc
[  0%] Built target genconfig
--
                 from /srv/home/jenkins/workspace/sof_generic_build/src/include/sof/lib/dma.h:20,
                 from /srv/home/jenkins/workspace/sof_generic_build/src/platform/intel/cavs/include/cavs/drivers/dw-dma.h:14,
                 from /srv/home/jenkins/workspace/sof_generic_build/src/platform/tigerlake/include/platform/drivers/dw-dma.h:13,
                 from /srv/home/jenkins/workspace/sof_generic_build/src/include/sof/drivers/dw-dma.h:12,
                 from /srv/home/jenkins/workspace/sof_generic_build/src/platform/intel/cavs/lib/dma.c:11:
/srv/home/jenkins/workspace/sof_generic_build/src/arch/xtensa/include/arch/atomic.h:69:3: error: #error No atomic methods available.
/srv/home/jenkins/workspace/sof_generic_build/src/arch/xtensa/include/arch/atomic.h:106:3: error: #error No atomic methods available.
[ 45%] Building C object CMakeFiles/sof.dir/src/platform/intel/cavs/lib/mem_window.c.o
[ 45%] Building C object CMakeFiles/sof.dir/src/platform/intel/cavs/platform.c.o
make[3]: *** [CMakeFiles/sof.dir/build.make:118: CMakeFiles/sof.dir/src/platform/intel/cavs/lib/dma.c.o] Error 2
```

Reply:
Hi Liam,

For gcc-build, there is no XCHAL_HAVE_EXCLUSIVE  and XCHAL_HAVE_S32C1I used, 

I wonder how it handles those cases before.

Reply:
That's a good point - then I think its best if we have for atomic.h and spinlock.h a set of functions for EXCLUSIVE and the existing set of functions for everything else. e.g.

```c
#if XCHAL_HAVE_EXCLUSIVE && CONFIG_XTENSA_EXCLUSIVE && __XCC__

/* All EXCLUSIVE methods here */

static inline int32_t arch_atomic_add(atomic_t *a, int32_t value)
{
	/*reference xtos : xipc_misc.h*/
	int32_t result = 0;
	int32_t current;

	while (!result) {
		current = XT_L32EX((int32_t *)a);
		result = current + value;
		XT_S32EX(result, (int32_t *)a);
		XT_GETEX(result);
	}
	return current;
}

#elif XCHAL_S32C1I

/* all S32C1I methods here */

static inline int32_t arch_atomic_add(atomic_t *a, int32_t value)
{
	int32_t result, current;

	__asm__ __volatile__(
		"1:     l32i    %1, %2, 0\n"
		"       wsr     %1, scompare1\n"
		"       add     %0, %1, %3\n"
		"       s32c1i  %0, %2, 0\n"
		"       bne     %0, %1, 1b\n"
		: "=&a" (result), "=&a" (current)
		: "a" (&a->value), "a" (value)
		: "memory");

	return current;
}

#else 

/* The xtensa config has no atomic method ISA support - we can use normal integer arithmetic iff UP */
#if CONFIG_NUM_CORES > 1
#error No atomic ISA for SMP configuration
#endif

/* integer arithmetic methods here */

#endif
```
This is simple do to in both files and wont break anything.